### PR TITLE
Adding script to setup watchman to synchronise into React Native apps

### PR DESCRIPTION
**Pull request by kraenhansen**
## What, How & Why?

This adds a `watch-from-dependency.js` script to help setup "watchman" to continuously monitor the repository and synchronise these changes into a React Native app's node_modules/realm directory.

I propose using "watchman" since it's a proven tool already used by the React Native toolchain and "rsync" since it's much more effective than a simple `rm -rf` followed by a `cp -r`.

## ☑️ ToDos
* [ ] 📝 Changelog entry
* [ ] 📝 `Compatibility` label is updated or copied from previous entry
* [ ] 🚦 Tests
* [x] 📝 Public documentation PR created or is not necessary
* [x] 💥 `Breaking` label has been applied or is not necessary

### CODE REVIEW

**Comment by kraenhansen** [comment]
I'm mostly asking reviews from all of you to get input on the wording in the file I've added to `contrib` as this should be easily understandable for any newcomer to the project.

**Review by blagoev — APPROVED** [review]
anything that speeds up development

File: scripts/watch-from-dependency.js [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+const TRIGGERS = [{
+  path: "src",
+  pattern: "src/*",
+}, {
+  path: "lib",
+  pattern: "lib/*",
+}, {
+  path: "react-native",
+  pattern: "react-native/*",
+}];
```
Comment (steffenagger) [inline scripts/watch-from-dependency.js:39]: Add `/types` to watchlist.

```suggestion
}, {
  path: "types",
  pattern: "types/*",
}];
```

**Review by steffenagger — COMMENTED** [review]
If I build the target RN app, while the watcher is running (in the background), I get the following error:
```sh
error: Build input file cannot be found: '/Users/steffen.agger/Projects/realm/try/RnTsApp/node_modules/realm/src/object-store/src/sync/remote_mongo_database.cpp' (in target 'RealmJS' from project 'Pods')
```

**Review by kneth — APPROVED** [review]

**Comment by kraenhansen** [comment]
@steffenagger I added [a note about this in the .md file](https://github.com/realm/realm-js/pull/3426/files#diff-9000472937dd862936992adb36faf2b04c8245ec907ba172b9cb7340b072cb6cR28-R31). You basically need to run `pod install` after the initial sync from the watcher. I might add something that actually runs `pod install` if a .cpp file is added or removed to the project 🤔  (at least at some point, that is).

**Comment by kraenhansen** [comment]
I did a major change of the code to use watchman subscriptions (via the Node.js SDK) instead of setting up persistent triggers (using the watchman CLI). This should be a lot more intuitive, making sure the files are only copied while the script is running.

File: .watchmanconfig [hunk 2]
```diff
@@ -0,0 +1,9 @@
+{
+  "ignore_dirs": [
+    "node_modules",
+    "react-native/node_modules",
+    "packages",
+    "integration-tests",
+    "tests"
+  ]
```
Comment (kraenhansen) [inline .watchmanconfig:8]: Listing these files makes the watch more performant.
We would need to remove these if we want to start reacting to changes in these directories.
Side note: Unfortunately this config doesn't understand a `**/node_modules/**` glob 😞

File: contrib/guide-watchman.md [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+```
+
+In the example above we're setting up a project-wide watch of our Realm JS repository and registering triggers on relevant directories (`/src`, `/lib` and `/react-native` at the time of writing) which will execute an "rsync" command to ultimately, incrementally (not a full copy every time) synchronize changes made our Realm JS repository into the app's `node_modules/realm` directory.
+
+## Stop watching
+
+To stop watching, simply delete the watch from watchman (from the realm-js project root directory)
+
+```
+watchman watch-del .
```
[the remark this answers is not in the document]
Comment (kraenhansen) [reply contrib/guide-watchman.md:40]: Makes sense - I updated to use a watchman subscription, which is removed when the process disconnects.

**Review by steffenagger — APPROVED** [review]
Looks good 👍 and works as intended (for the limited testing I've done).
Some suggestions added inline.

File: contrib/guide-watchman.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (kraenhansen) [reply contrib/guide-watchman.md:40]: As far as I understood it, having a watch created is not that big of an issue if there are no triggers or subscriptions added on it.

File: scripts/watch-from-dependency.js [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+      console.log(`🚀 Performing initial sync on ${resp.files.length} files`);
+    } else {
+      console.log(`🚀 Performing sync: ${resp.files.length} file(s) changed`);
+    }
+    cp.spawnSync("rsync", [
+      // "--verbose",
+      "--progress",
+      "--archive",
+      "--delete",
+      ...EXCLUDED_PATHS.map(p => ["--exclude", p]).flat(),
```
[the remark this answers is not in the document]
Comment (steffenagger) [reply scripts/watch-from-dependency.js:121]: Oh.. blind me. Maybe use reduce?
```js
EXCLUDED_PATHS.reduce((total, path) => [ ...total, '--exclude', path ], []);
```

[the remark this answers is not in the document]
Comment (steffenagger) [reply scripts/watch-from-dependency.js:121]: or use the combined exclude syntax: `--exclude={'path1','path2','path3'}` (which I just googled).
```js
`--exclude={${EXCLUDED_PATHS.map(p => `'${p}'`).join(',')}}`
```
Produces: `"--exclude={'react-native/node_modules','react-native/ios/Pods','vendor/realm-*'}"`

**Review by cesarvr — APPROVED** [review]
Looks good 👍

File: scripts/watch-from-dependency.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (kraenhansen) [reply scripts/watch-from-dependency.js:121]: One advantage of providing these directly in the args is that I don't have to worry about escaping special characters (in the weird case someone puts ' or } in their app's path 🙂

File: contrib/guide-watchman.md [hunk 5]
```diff
@@ -0,0 +1,41 @@
+# Guide: Setting up watchman to copy changes from this package to an app
+
+You're debugging an issue reproduced in a React Native app, which uses Realm JS. You want to interate fast, such that a change in the Realm JS source code gets reflected in the library used in the app, as fast as possible.
```
Comment (fronck) [inline contrib/guide-watchman.md:3]: *iterate

**Review by fronck — APPROVED** [review]
Small type in the readme.  Otherwise LGTM.

**Pull request merged**